File: volcanoPlot/volcanoCode_v1.1.R
library("dplyr")
library("tidyverse")
library("genefilter")
library("ggplot2")
library("grDevices")
library("ggrepel")

#Read in data frame with DE genes


volcanoPlot<-function(datainput, 
                      upColor="#bb0c00", 
                      downColor="blue2", 
                      unchangedColor="grey",
                      labelGenes=FALSE, # TRUE : Label genes
                      genes2label=NULL, # Deafult label top 10 genes, or custom number of genes, or gene names c("gene1", "gene2"...)
                      geneName.col="gene_name",
                      x.col="log2FoldChange", 
                      y.col="padj", 
                      plot.title="VolcanoPlot",
                      x.label="log2FoldChange",
                      y.label="-log10(padj)",
                      point.size=2,
                      transparency=1,
                      x.lower.limit=NULL,
                      x.Upper.limit=NULL,
                      x.break=NULL,
                      y.lower.limit=NULL,
                      y.Upper.limit=NULL,
                      y.break=NULL,
                      axis_text_size=5,
                      legend_title_size=12,
                      legend_text_size=8,
                      plot_title_size=12,
                      plot_title_face="bold",
                      label_size=5,
                      xy_axis_labels=14,
                      lfc_cut_off=1){
  
  # Data table formating
  datainput<-datainput[,c(geneName.col,x.col,y.col)]
  colnames(datainput)<-c("geneName","log2FoldChange","padj")
  datainput<-datainput[complete.cases(datainput),]
  DEstatus=ifelse(datainput["log2FoldChange"]>lfc_cut_off & datainput["padj"]<0.05,"Upregulated",
                  ifelse(datainput["log2FoldChange"] < (-lfc_cut_off) & datainput["padj"]<0.05,"Downregulated", "Not Significant"))
  DEstatus<-as.data.frame(DEstatus)
  colnames(DEstatus)<-"DEstatus"
  datainput<-cbind(datainput,DEstatus)
  x.col="log2FoldChange"
  y.col="padj"
  geneName.col="geneName"
  
  #datainput<-transform(datainput,DEstatus=ifelse(datainput[[x.col]]>1 & datainput[[y.col]]<0.05,"Upregulated",
  #                                    ifelse(datainput[[x.col]] < (-1) & datainput[[y.col]]<0.05,"Downregulated", "Not Significant")))
  #print(head(datainput,10))
  if (labelGenes==TRUE){
    #print(genes2label)
    if(is.null(genes2label)){
      message ("Top 10 genes with highest foldchange are labelled ")
      genes2label_list<- datainput %>% filter(.,abs(!!as.symbol(x.col)) > lfc_cut_off & (!!as.symbol(y.col)) < 0.05) %>% 
        arrange(desc(abs(!!as.symbol(x.col)))) %>% 
        slice_head(n=10) %>% 
        pull(geneName)
      print("Genes labeled in plot are:")
      print(paste(genes2label_list,sep=" "))
    }else if (is.vector(genes2label) & is.character(genes2label)){
      message ("Available genes from the genes2labels will be included in label")
      genes2label_list<-genes2label
      print("Genes labeled in plot are:")
      print(paste(genes2label_list,sep=" "))
  #}else if (!is.vector(genes2label & is.numeric(genes2label))){
    }else if (is.numeric(genes2label)){
      message (paste0("Top ", genes2label ," genes with highest foldchange are labelled"))
      genes2label_list<- datainput %>% filter(.,abs(!!as.symbol(x.col)) > lfc_cut_off & (!!as.symbol(y.col)) < 0.05) %>% 
        arrange(desc(abs(!!as.symbol(x.col)))) %>% 
        slice_head(n=genes2label) %>% 
        pull(geneName)
      print("Genes labeled in plot are:")
      print(paste(genes2label_list,sep=" "))
    }
  }else{
    genes2label_list=c()
  }
  
    
  CompletegeneList<-datainput[[geneName.col]]
  datainput$geneLabel <- ifelse(CompletegeneList %in%genes2label_list,CompletegeneList, NA )

  datainput<-transform(datainput,log10padj=-log10(datainput[[y.col]]))
  
  if(!is.null(x.lower.limit) & !is.null(x.Upper.limit) & !is.null(x.break)){
    x.axis.scale=sort(c(lfc_cut_off,-lfc_cut_off,seq(x.lower.limit,x.Upper.limit,x.break)))
    x_low_lim=x.lower.limit
    x_Up_lim=x.Upper.limit
  }else{
    x.axis.scale=sort(c(lfc_cut_off,-lfc_cut_off,seq(floor(min(datainput[x.col])*1.5), ceiling(max(datainput[x.col])*1.5), (((ceiling(max(datainput[x.col]))*1.5)-(floor((min(datainput[x.col]))*1.5)))/10))))
    x_low_lim=floor(min(x.axis.scale))
    x_Up_lim=ceiling(max(x.axis.scale))
  }
 # print(paste(floor(min(datainput[x.col])*1.5), ceiling(max(datainput[x.col])*1.5), (((ceiling(max(datainput[x.col]))*1.5)-(floor((min(datainput[x.col]))*1.5)))/10)))
#  print(paste(min(datainput["log10padj"])*1.15, max(datainput["log10padj"])*1.15, ((max(datainput["log10padj"])*1.15)-min(datainput["log10padj"])*1.15)/20))
  
  if(!is.null(y.lower.limit) & !is.null(y.Upper.limit) & !is.null(y.break)){
    y.axis.scale=seq(y.lower.limit,y.Upper.limit,y.break)
  }else{
    y.axis.scale=seq(min(datainput["log10padj"])*1.15, max(datainput["log10padj"])*1.15, ((max(datainput["log10padj"])*1.15)-min(datainput["log10padj"])*1.15)/20)
  }
  
  colourVector<-c(downColor, unchangedColor, upColor)
  names(colourVector)<-c("Downregulated", "Not Significant", "Upregulated")
  DEobsrvd<-sort(unique(datainput$DEstatus))
  colorIN<-colourVector[DEobsrvd]
  names(colorIN)<-NULL
  
 # print(y.axis.scale)
 
  ####Plot Code
  ggplot(data = datainput, aes_string(x = x.col, y = "log10padj", col = "DEstatus", label = "geneLabel")) +
    geom_vline(xintercept = c(-lfc_cut_off, lfc_cut_off), col = "blue", linetype = 'dashed', alpha=0.5) +
    geom_hline(yintercept = -log10(0.05), col = "blue", linetype = 'dashed', alpha=0.5) + 
    geom_point(size = point.size, alpha=transparency) + 
    scale_color_manual(values = colorIN,
                       labels = DEobsrvd) +
    #scale_color_manual(values = c(downColor, unchangedColor, upColor), # to set the colours of our variable  
    #                   labels = c("Downregulated", "Not significant", "Upregulated")) + # to set the labels in case we want to overwrite the categories from the dataframe (UP, DOWN, NO)
    coord_cartesian(ylim = c(0, ceiling(max(y.axis.scale))), xlim = c(x_low_lim, x_Up_lim)) + # since some genes can have minuslog10padj of inf, we set these limits
    #coord_cartesian(ylim = c(0, 80), xlim = c(-10,10)) +
    labs(x = x.label, y = y.label) + 
    scale_x_continuous(breaks = x.axis.scale) + # to customise the breaks in the x axis
    ggtitle(plot.title) + # Plot title 
    geom_text_repel(max.overlaps = Inf,size=label_size) + # To show all labels 
    theme_classic() +
    theme(text = element_text(size=rel(axis_text_size)), #change fontsize of axis ticks and label
          legend.key.size = unit(0.5, 'cm'), #change legend key size
          legend.key.height = unit(0.5, 'cm'), #change legend key height
          legend.key.width = unit(0.5, 'cm'), #change legend key width
          legend.title = element_text(size=legend_title_size), #change legend title font size
          legend.text = element_text(size=legend_text_size),
          title =element_text(size=plot_title_size, face=plot_title_face), #change legend text font size)
          axis.title = element_text(family = "Helvetica", size = xy_axis_labels, colour = "black"))
}
#)

#df<-read.csv("../Sanja/DE_results_01/CLuster_2_nicdFrac_vs_cntFrac.csv")

# #volcanoPlot(df,
#             labelGenes=TRUE,
#             geneName.col="geneID",
#             x.col="avg_log2FC", 
#             y.col="p_val_adj", 
#             plot.title="VolcanoPlot : MSC2",
#             x.label="log2FoldChange",
#             y.label="-log10(padj)",
#             x.Upper.limit=2.5,
#             x.lower.limit=-1.5,
#             x.break=2,
#             y.lower.limit=0,
#             y.Upper.limit=100,
#             y.break=50,
#             point.size=2,
#             transparency=0.5,
#             legend_text_size=12,
#             label_size=7,
#             xy_axis_labels=18,
#             genes2label=5)

File: volcanoPlot/volcanoCode_v1.1.py
import pandas as pd
import matplotlib.pyplot as plt

def plot_volcano(file_path, id_col='gene_name', x_col='log2FoldChange', y_col='padj',
                 up_color='red', down_color='green', neutral_color='grey', label_genes=False,
                 genes_to_label=None, plot_title='Volcano plot', point_size=50, alpha=0.6,
                 x_lim=None, x_break=None, y_lim=None, y_break=None,
                 x_label_size=12, x_title_size=14, y_label_size=12, y_title_size=14,
                 label_size=10, log2FC_threshold=(1, -1), padj_threshold=0.05):
    """
    Generate a volcano plot for gene expression data.

    Parameters:
    - file_path: Path to the CSV file containing the data.
    - id_col: Column name for gene identifiers.
    - x_col: Column name for log2 fold change values.
    - y_col: Column name for adjusted p-values.
    - up_color, down_color, neutral_color: Colors for upregulated, downregulated, and not significant points.
    - label_genes: Boolean to indicate if genes should be labeled.
    - genes_to_label: List of gene names to label or number to label the top changing genes.
    - plot_title: Title of the plot.
    - point_size: Size of the points in the plot.
    - alpha: Transparency of the points.
    - x_lim, y_lim: Tuple for x and y axis limits (min, max).
    - x_break, y_break: Break size on the x and y axis.
    - x_label_size, y_label_size: Font size for axis labels.
    - x_title_size, y_title_size: Font size for axis titles.
    - label_size: Font size for data point labels.
    - log2FC_threshold: Thresholds for significant log2 fold change.
    - padj_threshold: Threshold for significant adjusted p-value.
    """

    # Load data
    data = pd.read_csv(file_path)

    # Determine colors for each data point based on significance thresholds
    conditions = [
        (data[x_col] >= log2FC_threshold[0]) & (data[y_col] <= padj_threshold),
        (data[x_col] <= log2FC_threshold[1]) & (data[y_col] <= padj_threshold)
    ]
    choices = [up_color, down_color]
    data['color'] = pd.Series(neutral_color, index=data.index)
    data['color'] = pd.Series(np.select(conditions, choices, default=neutral_color), index=data.index)

    # Plotting
    plt.figure(figsize=(10, 8))
    plt.scatter(data[x_col], -np.log10(data[y_col]), c=data['color'], alpha=alpha, s=point_size)

    # Label genes if required
    if label_genes:
        if genes_to_label is None or type(genes_to_label) == int:
            top_genes = data.nlargest(genes_to_label or 10, key=lambda x: abs(x[x_col]))
        else:
            top_genes = data[data[id_col].isin(genes_to_label)]

        for _, row in top_genes.iterrows():
            plt.text(row[x_col], -np.log10(row[y_col]), row[id_col], fontsize=label_size)

    # Draw significance threshold lines
    plt.axhline(-np.log10(padj_threshold), color='grey', linestyle='dashed')
    plt.axvline(log2FC_threshold[0], color='grey', linestyle='dashed')
    plt.axvline(log2FC_threshold[1], color='grey', linestyle='dashed')

    # Set limits and labels
    plt.xlim(x_lim or (data[x_col].min() - x_break if x_break else None, data[x_col].max() + x_break if x_break else None))
    plt.ylim(y_lim or (-np.log10(data[y_col].max()) - y_break if y_break else None, 0 + y_break if y_break else None))
    plt.xlabel('Log2 Fold Change', fontsize=x_title_size)
    plt.ylabel('-Log10 adjusted P-value', fontsize=y_title_size)
    plt.title(plot_title)

    # Show plot
    plt.show()

# Example usage
plot_volcano('path_to_your_data.csv')
